src/colmap/controllers/feature_extraction.cc

```diff
@@ -423,10 +417,59 @@ class FeatureExtractorController : public Thread {
     }
 
     writer_ = std::make_unique<FeatureWriterThread>(
-        image_reader_.NumImages(), &database_, writer_queue_.get());
+        database_.NumImages(), &database_, writer_queue_.get());
   }
 
  private:
+  ImageData ReadImageData(const Image& image) {
+    ImageData image_data;
+    image_data.image = image;
+    image_data.camera = database_.ReadCamera(image.CameraId());
+
+    const bool exists_keypoints = database_.ExistsKeypoints(image.ImageId());
+    const bool exists_descriptors =
+        database_.ExistsDescriptors(image.ImageId());
+    if (exists_keypoints && exists_descriptors) {
+      image_data.status = FeatureExtractionStatus::IMAGE_EXISTS;
+      return image_data;
+    }
+
+    // Construct full image path and load bitmap
+    const std::string image_path = JoinPaths(image_root_path_, image.Name());
+    if (!image_data.bitmap.Read(image_path, false)) {
+      LOG(ERROR) << image.Name() << " Failed to read the image file format.";
+      image_data.status = FeatureExtractionStatus::FAILURE;
+      return image_data;
+    }
+
+    // Load mask if path is provided
+    if (!mask_root_path_.empty()) {
+      std::string mask_path = JoinPaths(mask_root_path_, image.Name() + ".png");
+      bool exists_mask = true;
+      if (!ExistsFile(mask_path)) {
+        exists_mask = false;
+        // Try without .png extension if original image is .png
+        if (HasFileExtension(image.Name(), ".png")) {
+          std::string alt_mask_path = JoinPaths(mask_root_path_, image.Name());
+          if (ExistsFile(alt_mask_path)) {
+            mask_path = std::move(alt_mask_path);
+            exists_mask = true;
+          }
+        }
+      }
+      if (exists_mask) {
+        if (!image_data.mask.Read(mask_path, false)) {
+          LOG(ERROR) << image.Name() << " Failed to read the mask file";
+          image_data.mask.Deallocate();  // Skip the mask but not the image!
```

Comment:
Previously, no features would be extracted for an image when the mask cannot be found or read. This sounds wrong to me.